internal/wire/msg_body.go

```diff
@@ -150,34 +151,42 @@ func getChecksum(data []byte) (uint32, error) {
 	return binary.LittleEndian.Uint32(data[n-crc32.Size:]), nil
 }
 
-func containsValidChecksum(header *MsgHeader, body []byte) error {
+// validateChecksum calculates checksum of the message (header + body)
+// and compares it with the checksum from the last bytes of the message.
+// If the flag bit for checksum presence is not set or the checksum is valid, it returns nil.
+// If the checksum is invalid, it returns an error.
+//
+// TODO The callers of checksum validation should be closer to OP_MSG handling: https://github.com/FerretDB/FerretDB/issues/2690
+func validateChecksum(header *MsgHeader, body []byte) error {
 	if len(body) < kFlagBitSize {
 		return lazyerrors.New("Message contains illegal flags value")
 	}
 
 	flagBit := OpMsgFlags(binary.LittleEndian.Uint32(body[:kFlagBitSize]))
-	if flagBit.FlagSet(OpMsgChecksumPresent) {
-		want, err := getChecksum(body)
-		if err != nil {
-			lazyerrors.Error(err)
-		}
+	if !flagBit.FlagSet(OpMsgChecksumPresent) {
+		return nil
+	}
 
-		// https://datatracker.ietf.org/doc/html/rfc4960#appendix-B
-		hasher := crc32.New(crc32.MakeTable(crc32.Castagnoli))
+	want, err := getChecksum(body)
+	if err != nil {
+		return lazyerrors.Error(err)
+	}
 
-		if err := binary.Write(hasher, binary.LittleEndian, header); err != nil {
-			return lazyerrors.Error(err)
-		}
+	// https://datatracker.ietf.org/doc/html/rfc4960#appendix-B
+	hasher := crc32.New(crc32.MakeTable(crc32.Castagnoli))
 
-		offset := len(body) - crc32.Size
-		if err := binary.Write(hasher, binary.LittleEndian, body[:offset]); err != nil {
-			return lazyerrors.Error(err)
-		}
+	if err := binary.Write(hasher, binary.LittleEndian, header); err != nil {
+		return lazyerrors.Error(err)
+	}
 
-		got := hasher.Sum32()
-		if want != got {
-			return lazyerrors.New("OP_MSG checksum does not match contents.")
-		}
+	offset := len(body) - crc32.Size
+	if err := binary.Write(hasher, binary.LittleEndian, body[:offset]); err != nil {
+		return lazyerrors.Error(err)
+	}
```

Comment:
> To write the checksum, we need to modify the header by summing four bytes in the message length field for the checksum.

Does the quote from https://github.com/FerretDB/FerretDB/issues/2690 refer above two statements? Just checking :hugs:

Reply:
The above two statements are to calculate the checksum. To calculate checksum we need the whole header plus the whole body except the last four bytes.